File: samples/conference/registration-bc/src/main/java/com/microsoft/conference/registration/web/Controllers/OrderController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.Web.Mvc;
using Conference.Common;
using ECommon.IO;
using ENode.Commanding;
using Registration.Commands;
using Registration.Commands.SeatAssignments;
using Registration.ReadModel;
using Registration.Web.Models;

namespace Registration.Web.Controllers
{
    public class OrderController : ConferenceTenantController
    {
        private readonly ICommandService _commandService;

        public OrderController(ICommandService commandService, IConferenceQueryService conferenceQueryService, IOrderQueryService orderQueryService)
            : base(conferenceQueryService, orderQueryService)
        {
            _commandService = commandService;
        }

        [HttpGet]
        public ActionResult Display(Guid orderId)
        {
            var order = OrderQueryService.FindOrder(orderId);
            if (order == null)
            {
                return RedirectToAction("Find", new { conferenceCode = this.ConferenceCode });
            }
            return View(order);
        }

        [HttpGet]
        [OutputCache(Duration = 0, NoStore = true)]
        public RedirectToRouteResult AssignSeatsForOrder(Guid orderId)
        {
            var order = OrderQueryService.FindOrder(orderId);
            if (order == null)
            {
                return RedirectToAction("Display", new { orderId });
            }

            return RedirectToAction("AssignSeats", new { assignmentsId = order.OrderId });
        }

        [HttpGet]
        [OutputCache(Duration = 0, NoStore = true)]
        public ActionResult AssignSeats(Guid orderId)
        {
            var assignments = OrderQueryService.FindOrderSeatAssignments(orderId);
            if (assignments == null)
            {
                return RedirectToAction("Find", new { conferenceCode = this.ConferenceCode });
            }
            return View(new OrderSeatsViewModel { OrderId = orderId, SeatAssignments = assignments });
        }

        [HttpPost]
        public ActionResult AssignSeats(Guid orderId, List<OrderSeatAssignment> seatAssignments)
        {
            if (!seatAssignments.Any())
            {
                return RedirectToAction("Display", new { orderId = orderId });
            }

            var assignmentsId = seatAssignments[0].AssignmentsId;
            var unassignedCommands = seatAssignments
                .Where(x => string.IsNullOrWhiteSpace(x.AttendeeEmail))
                .Select(x => (ICommand)new UnassignSeat(assignmentsId) { Position = x.Position });
            var assignedCommands = seatAssignments
                .Where(x => !string.IsNullOrWhiteSpace(x.AttendeeEmail))
                .Select(x => new AssignSeat(assignmentsId)
                {
                    Position = x.Position,
                    PersonalInfo = new PersonalInfo
                    {
                        Email = x.AttendeeEmail,
                        FirstName = x.AttendeeFirstName,
                        LastName = x.AttendeeLastName
                    }
                });

            var commands = assignedCommands.Union(unassignedCommands).ToList();
            foreach (var command in commands)
            {
                SendCommandAsync(command);
            }

            return RedirectToAction("Display", new { orderId = orderId });
        }

        [HttpGet]
        public ActionResult Find()
        {
            return View();
        }

        [HttpPost]
        public ActionResult Find(string email, string accessCode)
        {
            var orderId = OrderQueryService.LocateOrder(email, accessCode);

            if (!orderId.HasValue)
            {
                // TODO: 404?
                return RedirectToAction("Find", new { conferenceCode = this.ConferenceCode });
            }

            return RedirectToAction("Display", new { conferenceCode = this.ConferenceCode, orderId = orderId.Value });
        }

        /// <summary>异步发送给定的命令
        /// </summary>
        /// <param name="command"></param>
        /// <param name="millisecondsDelay"></param>
        /// <returns></returns>
        private Task<AsyncTaskResult> SendCommandAsync(ICommand command, int millisecondsDelay = 5000)
        {
            return _commandService.SendAsync(command).TimeoutAfter(millisecondsDelay);
        }
    }
}

File: samples/conference/registration-bc/src/main/java/com/microsoft/conference/registration/controller/OrderController.java
package com.microsoft.conference.registration.controller;

import com.google.common.base.Strings;
import com.microsoft.conference.common.ActionResult;
import com.microsoft.conference.common.ErrCode;
import com.microsoft.conference.common.registration.commands.PersonalInfo;
import com.microsoft.conference.common.registration.commands.seatassignment.AssignSeat;
import com.microsoft.conference.common.registration.commands.seatassignment.UnassignSeat;
import com.microsoft.conference.registration.readmodel.service.OrderQueryService;
import com.microsoft.conference.registration.readmodel.service.OrderSeatAssignmentVO;
import com.microsoft.conference.registration.readmodel.service.OrderVO;
import org.enodeframework.commanding.ICommand;
import org.enodeframework.commanding.ICommandService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.stream.Collectors;

@RestController
public class OrderController {

    @Autowired
    private ICommandService commandService;

    @Autowired
    private OrderQueryService orderQueryService;

    @GetMapping("orders/{id}")
    public ActionResult<OrderVO> display(@PathVariable("id") String orderId) {
        OrderVO order = orderQueryService.findOrder(orderId);
        return view(order);
    }

    @GetMapping("orders")
    public ActionResult<String> find(String email, String accessCode) {
        String orderId = orderQueryService.locateOrder(email, accessCode);
        return view(orderId);
    }

    @GetMapping("seatassignments/{id}")
    public ActionResult<OrderSeatsViewModel> assignSeats(@PathVariable("id") String orderId) {
        List<OrderSeatAssignmentVO> assignments = orderQueryService.findOrderSeatAssignments(orderId);
        OrderSeatsViewModel viewModel = new OrderSeatsViewModel();
        viewModel.setOrderId(orderId);
        viewModel.setSeatAssignments(assignments);
        return view(viewModel);
    }

    @PostMapping("seatassignments")
    public ActionResult<String> assignSeats(OrderSeatsViewModel viewModel) {
        List<OrderSeatAssignmentVO> seatAssignments = viewModel.getSeatAssignments();
        if (seatAssignments.isEmpty()) {
            return ActionResult.error(ErrCode.SYSTEM_ERROR, "" + viewModel.getOrderId());
        }
        String assignmentsId = seatAssignments.get(0).getAssignmentsId();
        List<ICommand> unassignedCommands = seatAssignments.stream().filter(x -> Strings.isNullOrEmpty(x.getAttendeeEmail()))
                .map(y -> {
                    UnassignSeat unassignSeat = new UnassignSeat(assignmentsId);
                    unassignSeat.setPosition(y.getPosition());
                    return (ICommand) unassignSeat;
                }).collect(Collectors.toList());
        List<ICommand> assignedCommands = seatAssignments.stream().filter(x -> !Strings.isNullOrEmpty(x.getAttendeeEmail()))
                .map(y -> {
                    AssignSeat unassignSeat = new AssignSeat(assignmentsId);
                    unassignSeat.setPosition(y.getPosition());
                    PersonalInfo personalInfo = new PersonalInfo();
                    personalInfo.setEmail(y.getAttendeeEmail());
                    personalInfo.setFirstName(y.getAttendeeFirstName());
                    personalInfo.setLastName(y.getAttendeeLastName());
                    unassignSeat.setPersonalInfo(personalInfo);
                    return (ICommand) unassignSeat;
                }).collect(Collectors.toList());
        assignedCommands.addAll(unassignedCommands);
        for (ICommand command : assignedCommands) {
            sendCommandAsync(command);
        }
        return ActionResult.empty();
    }


    private <T> ActionResult<T> view(T objects) {
        return new ActionResult<>(objects);
    }

    private void sendCommandAsync(ICommand command) {
        commandService.sendAsync(command);
    }
}